packages/cli-platform-android/src/config/findPackageClassName.ts

```diff
@@ -29,6 +29,35 @@ export function getMainActivityFiles(
 }
 
 export default function getPackageClassName(folder: string) {
+  /*
+    When module contains `expo-module.config.json` we return null
+    because expo modules follow other practices and don't implement
+    ReactPackage/TurboReactPackage directly, so it doesn't make sense
+    to scan and read hundreds of files to get package class name.
+
+    Exception is `expo` package itself which contains `expo-module.config.json`
+    and implements `ReactPackage/TurboReactPackage` inside `ExpoModulesPackage.kt`.
+
+    Following logic is done due to performance optimization.
+  */
+
+  if (
+    fs.existsSync(path.join(folder, '..', 'expo-module.config.json')) &&
+    !fs.existsSync(
+      path.join(
+        folder,
+        'src',
+        'main',
+        'java',
+        'expo',
+        'modules',
+        'ExpoModulesPackage.kt',
```

Comment:
looks good to me. we had the similar logic at https://github.com/expo/expo/blob/08cba033007a260318fc438ab1d9f872d4ebac2f/packages/expo-modules-autolinking/src/reactNativeConfig/androidResolver.ts#L113-L125.

not sure if that's possible for not hardcoding `ExpoModulesPackage.kt` here. that's not a public api.

a side note. when implementing autolinking at expo, i feel that on android we waste much work on glob/file parsing. maybe at some point we could let developers to specify java class names or package names in **react-native.config.js**

Reply:
> not sure if that's possible for not hardcoding ExpoModulesPackage.kt

Is there anything else that we could check to ensure that module is `expo` package? 

> maybe at some point we could let developers to specify java class names or package names in react-native.config.js

IMO asking every library maintainer do add this value might be hard, we consider it some time ago. https://github.com/react-native-community/cli/pull/2384 this change should be already fine for majority of cases.

Reply:
> Is there anything else that we could check to ensure that package is `expo` package?

in our implementation, we try to match `*Package.{java,kt}` and check whether the file extends `ReactPackage`/`TurboReactPackage` before the `expo-modules.config.json` skip path. the `expo` package will matched by the `*Package.{java,kt}` check.
not sure if implementation helps for rnc-cli though. https://github.com/expo/expo/blob/08cba033007a260318fc438ab1d9f872d4ebac2f/packages/expo-modules-autolinking/src/reactNativeConfig/androidResolver.ts#L113-L125

alternative, to exclude the `expo` package, is it possible to pass packageName into the function? maybe we can also skip when `packageName === 'expo'`
 
> IMO asking every library maintainer do add this value might be hard, we consider it some time ago. #2384 this change should be already fine for majority of cases.

as long as one lib adding this value for optimization, it could help to reduce the whole app autolinking time. but that's fine if's that's being discussed already.

Reply:
> let's remember to always have a fallback across implementations because there's no official guideline to have implments ReactPackage/TurboReactPackage inside file with *Package.kt/java file (and also there are some community libraries that don't follow this rule).

yep we have the fallback as https://github.com/expo/expo/blob/08cba033007a260318fc438ab1d9f872d4ebac2f/packages/expo-modules-autolinking/src/reactNativeConfig/androidResolver.ts#L123-L129. thanks for sharing this!